logs/querysample/tracing.go

```diff
@@ -0,0 +1,152 @@
+	fpBin := make([]byte, 8)
+	binary.BigEndian.PutUint64(fpBin, fp)
+
+	return fmt.Sprintf(
+		"%s/databases/%s/queries/%s/samples/%d?role=%s",
+		grant.Config.ServerURL,
+		sample.Database,
+		hex.EncodeToString(fpBin),
+		sample.OccurredAt.Unix(),
+		sample.Username,
+	)
+}
+
+func parseTags(query string) map[string]string {
+	tokens, err := pg_query.Scan(query)
+	if err != nil {
+		return nil
+	}
+
+	tags := make(map[string]string)
+	for _, token := range tokens.GetTokens() {
+		var comment string
+		if token.Token == pg_query.Token_SQL_COMMENT {
+			comment = query[token.Start+2 : token.End]
+		} else if token.Token == pg_query.Token_C_COMMENT {
+			comment = query[token.Start+2 : token.End-2]
+		} else {
+			continue
+		}
+		for _, part := range strings.Split(strings.TrimSpace(comment), ",") {
+			if strings.Contains(part, "=") {
+				keyAndValue := strings.SplitN(part, "=", 2)
+				tags[keyAndValue[0]] = keyAndValue[1]
+				// TODO: Handle sqlcommenter's URL encoding logic, and quoting with single quotes
+			} else if strings.Contains(part, ":") {
+				keyAndValue := strings.SplitN(part, ":", 2)
+				tags[keyAndValue[0]] = keyAndValue[1]
+			} // TODO: Support other formats
+		}
+	}
+	return tags
+}
+
+func initProvider(ctx context.Context, endpoint string) (*sdktrace.TracerProvider, func(context.Context) error, error) {
+	res, err := resource.New(ctx,
+		resource.WithAttributes(
+			// the service name used to display traces in backends
+			semconv.ServiceName("Postgres (pganalyze)"),
+		),
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create resource: %w", err)
+	}
+
+	url, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to parse endpoint URL: %w", err)
+	}
+	opts := []otlptracehttp.Option{otlptracehttp.WithEndpoint(url.Host)}
+	switch strings.ToLower(url.Scheme) {
+	case "http":
+		opts = append(opts, otlptracehttp.WithInsecure())
+	case "https":
+		// Secure is default
+	default:
+		// TODO: Should we support "grpc" here? (it seems we're not really avoiding to import its code anyway?)
+		return nil, nil, fmt.Errorf("unsupported protocol: %s", url.Scheme)
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second)
+	defer cancel()
+	traceExporter, err := otlptracehttp.New(ctx, opts...)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create trace exporter: %w", err)
+	}
+
+	bsp := sdktrace.NewBatchSpanProcessor(traceExporter)
+	tracerProvider := sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sdktrace.AlwaysSample()),
+		sdktrace.WithResource(res),
+		sdktrace.WithSpanProcessor(bsp),
+	)
+
+	return tracerProvider, tracerProvider.Shutdown, nil
+}
+
+func ReportQuerySamplesAsTraceSpans(ctx context.Context, server *state.Server, grant state.GrantLogs, samples []state.PostgresQuerySample) {
+	endpoint := server.Config.OtelExporterOtlpEndpoint
+	if endpoint == "" {
+		return
+	}
+
+	tracerProvider, shutdown, err := initProvider(ctx, server.Config.OtelExporterOtlpEndpoint)
```

Comment:
We're re-initializing this for every log snapshot, right? I'm not familiar with the SDK, so not sure if that's the recommended approach, but it seems like it'd be more expensive than initializing once per collector startup (if nothing else, in terms of keeping the TCP connection open). But reorganizing that could get a bit messy, so we can start with this.

Reply:
Yeah, good catch. As you can imagine its not the recommended approach to reinitialize it. The overall idea is that you initialize this once and then your whole app uses the same trace provider via some global helpers.

That said, it seemed easiest for now to do it like this to keep this code in one place, and to ensure all spans are sent when the snapshot is processed (i.e. not with a delay) - I'll leave a note to address this in a future revision (I do think it makes sense to centralize this, just need to spend more time testing that).